nvflare/app_opt/job_launcher/k8s_launcher.py

```diff
@@ -0,0 +1,263 @@
+            c = Configuration().get_default_copy()
+        except AttributeError:
+            c = Configuration()
+            c.assert_hostname = False
+        Configuration.set_default(c)
+        self.core_v1 = core_v1_api.CoreV1Api()
+        self.namespace = namespace
+
+        self.job_handle = None
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+    def launch_job(self, launch_data: dict, fl_ctx: FLContext) -> JobHandleSpec:
+
+        workspace_obj: Workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        args = fl_ctx.get_prop(FLContextKey.ARGS)
+        client = fl_ctx.get_prop(FLContextKey.SITE_OBJ)
+        job_id = launch_data.get(JobConstants.JOB_ID)
+        server_config = fl_ctx.get_prop(FLContextKey.SERVER_CONFIG)
+        if not server_config:
+            raise RuntimeError(f"missing {FLContextKey.SERVER_CONFIG} in FL context")
+        service = server_config[0].get("service", {})
+        if not isinstance(service, dict):
+            raise RuntimeError(f"expect server config data to be dict but got {type(service)}")
+
+        self.logger.info(f"K8sJobLauncher start to launch job: {job_id} for client: {client.client_name}")
+        job_image = launch_data.get(JobConstants.JOB_IMAGE)
+        self.logger.info(f"launch job use image: {job_image}")
+        job_config = {
+            "name": job_id,
+            "image": job_image,
+            "container_name": f"container-{job_id}",
+            "volume_mount_list": [{"name": self.workspace, "mountPath": self.mount_path}],
+            "volume_list": [{"name": self.workspace, "hostPath": {"path": self.root_hostpath, "type": "Directory"}}],
+            "module_args": {
+                "-m": args.workspace,
+                "-w": (workspace_obj.get_startup_kit_dir()),
+                "-t": client.token,
+                "-d": client.ssid,
+                "-n": job_id,
+                "-c": client.client_name,
+                "-p": "tcp://parent-pod:8004",
+                "-g": service.get("target"),
+                "-scheme": service.get("scheme", "grpc"),
+                "-s": "fed_client.json",
+            },
+            "set_list": args.set,
+        }
+
+        self.logger.info(f"launch job with k8s_launcher. Job_id:{job_id}")
+
+        job_handle = K8sJobHandle(job_id, self.core_v1, job_config, namespace=self.namespace, timeout=self.timeout)
+        try:
+            self.core_v1.create_namespaced_pod(body=job_handle.get_manifest(), namespace=self.namespace)
+            if job_handle.enter_states([JobState.RUNNING], timeout=self.timeout):
+                return job_handle
+            else:
+                job_handle.terminate()
+                return None
+        except ApiException as e:
+            job_handle.terminate()
+            return None
+
+    def can_launch(self, launch_data: dict) -> bool:
+        job_image = launch_data.get(JobConstants.JOB_IMAGE)
+        if job_image in self.supported_images:
```

Comment:
The job image is job specific, and there can be any number of jobs, how can self.supported_images cover all of these?

Reply:
removed supported_images.